File: core/utils/file.py
import logging
import re
from pathlib import Path

from httpx import Client, Response
from pytonapi.schema.nft import ImagePreview

from core.constants import STATIC_PATH

client = Client()

logger = logging.getLogger(__name__)


CONTENT_DISPOSITION_FILENAME_REGEX = re.compile(r'filename="(.+)"')


def get_filename_from_content_disposition(response: Response) -> str | None:
    """
    Extract filename from Content-Disposition header.
    """
    content_disposition = response.headers.get("Content-Disposition")
    match = CONTENT_DISPOSITION_FILENAME_REGEX.search(content_disposition)
    if match:
        return match.group(1)
    return None


def guess_file_extension(response: Response) -> str | None:
    """
    Guess the file extension from the response content type.
    """
    content_type = response.headers.get("Content-Type")
    if content_type:
        return content_type.split("/")[-1]

    if filename := get_filename_from_content_disposition(response):
        return filename.split(".")[-1]

    return None


def download_media(
    url: str,
    name: str,
    subdirectory: str | Path | None = None,
    default_extension: str = ".webp",
) -> Path | None:
    """
    Download media from URL.

    :param url: URL to download media from.
    :param name: Name of the file that will be used. Should not include the extension.
    :param subdirectory: Subdirectory to save the file in.
    :param default_extension: Default extension to use if the extension cannot be guessed.

    :return: Path to the downloaded file if file was downloaded successfully, None otherwise.
    """
    root_path = STATIC_PATH / (subdirectory or "")

    try:
        response = client.get(url)
    except:  # noqa: E722
        logger.exception("Unable to download media")
        return None

    file_name = f"{name}.{guess_file_extension(response) or default_extension}"
    full_path = root_path / file_name
    with open(full_path, "wb") as file:
        file.write(response.content)

    return full_path


def pick_best_preview(previews: list[ImagePreview]) -> ImagePreview:
    """
    Pick the best image preview from a list of previews.
    """
    try:
        return max(
            previews,
            key=lambda preview: preview.resolution.split("x")[0]
            * preview.resolution.split("x")[1],
        )
    except (TypeError, ValueError):
        logger.warning("Could not pick the best preview. Returning the last one")
        return previews[-1]


def clean_old_versions(
    path: Path,
    prefix: str,
    current_file: str,
) -> None:
    """
    Clean old versions of files in a path

    :param path: Path to the directory containing the files.
    :param prefix: Prefix of the files to clean.
    :param current_file: Current file name.
    """
    for file in path.glob(f"{prefix}*"):
        if file.name != current_file:
            file.unlink()

File: lets-go/internal/utils/file.go
package utils

import (
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Define regular expressions
var contentDispositionFilenameRegex = regexp.MustCompile(`filename="(.+)"`)

// getFilenameFromContentDisposition extracts the filename from the Content-Disposition header
func getFilenameFromContentDisposition(headers http.Header) (string, error) {
	contentDisposition := headers.Get("Content-Disposition")
	if contentDisposition == "" {
		return "", errors.New("Content-Disposition header not found")
	}

	match := contentDispositionFilenameRegex.FindStringSubmatch(contentDisposition)
	if len(match) > 1 {
		return match[1], nil
	}
	return "", errors.New("Filename not found in Content-Disposition header")
}

// guessFileExtension guesses the file extension from the Content-Type header or the Content-Disposition header
func guessFileExtension(resp *http.Response) (string, error) {
	// Try to get from Content-Type header
	contentType := resp.Header.Get("Content-Type")
	if contentType != "" {
		parts := strings.Split(contentType, "/")
		if len(parts) > 1 {
			return parts[len(parts)-1], nil
		}
	}

	// Fallback to Content-Disposition header
	filename, err := getFilenameFromContentDisposition(resp.Header)
	if err == nil && strings.Contains(filename, ".") {
		ext := filepath.Ext(filename)
		if ext != "" {
			return ext[1:], nil // Remove the leading dot
		}
	}

	return "", errors.New("Unable to guess file extension")
}

// downloadMedia downloads a file from the given URL and saves it to the specified path on disk
func DownloadMedia(url, name, subdirectory, defaultExtension string, staticPath string) (string, error) {
	// Ensure default extension starts with a dot
	if !strings.HasPrefix(defaultExtension, ".") {
		defaultExtension = "." + defaultExtension
	}

	// Resolve the full path
	rootPath := staticPath
	if subdirectory != "" {
		rootPath = filepath.Join(staticPath, subdirectory)
	}

	// Create the directories if necessary
	if err := os.MkdirAll(rootPath, os.ModePerm); err != nil {
		return "", err
	}

	// Make the HTTP GET request
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	// Guess the file extension
	extension, err := guessFileExtension(resp)
	if err != nil {
		extension = defaultExtension
	}

	// Create the destination file name and path
	fileName := name + extension
	fullPath := filepath.Join(rootPath, fileName)

	// Create and write the file
	file, err := os.Create(fullPath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	_, err = io.Copy(file, resp.Body)
	if err != nil {
		return "", err
	}

	return fullPath, nil
}

// ImagePreview represents an image preview with a resolution
type ImagePreview struct {
	Resolution string
}

// parseResolution parses a resolution string (e.g., "1920x1080") into width and height
func parseResolution(resolution string) (int, int, error) {
	parts := strings.Split(resolution, "x")
	if len(parts) != 2 {
		return 0, 0, errors.New("Invalid resolution format")
	}

	width, err := strconv.Atoi(parts[0])
	if err != nil {
		return 0, 0, err
	}

	height, err := strconv.Atoi(parts[1])
	if err != nil {
		return 0, 0, err
	}

	return width, height, nil
}

// pickBestPreview selects the best image preview based on its resolution
func PickBestPreview(previews []ImagePreview) (ImagePreview, error) {
	if len(previews) == 0 {
		return ImagePreview{}, errors.New("No previews available")
	}

	sort.Slice(previews, func(i, j int) bool {
		w1, h1, err1 := parseResolution(previews[i].Resolution)
		w2, h2, err2 := parseResolution(previews[j].Resolution)

		if err1 != nil || err2 != nil {
			return false
		}

		return (w1 * h1) < (w2 * h2)
	})

	// Return the best (last after sort)
	return previews[len(previews)-1], nil
}

// cleanOldVersions removes old files in a directory with a specific prefix, excluding the current file
func CleanOldVersions(path, prefix, currentFile string) error {
	files, err := os.ReadDir(path)
	if err != nil {
		return err
	}

	for _, file := range files {
		if strings.HasPrefix(file.Name(), prefix) && file.Name() != currentFile {
			err = os.Remove(filepath.Join(path, file.Name()))
			if err != nil {
				return err
			}
		}
	}

	return nil
}